blockdag/utxoio.go

```diff
@@ -217,97 +217,107 @@ func serializeMultiset(w io.Writer, ms *ecc.Multiset) error {
 }
 
 // serializeUTXO serializes a utxo entry-outpoint pair
-func serializeUTXO(w io.Writer, entry *UTXOEntry, outpoint *wire.Outpoint) error {
-	serializedOutpoint := *outpointKey(*outpoint)
-	err := wire.WriteVarInt(w, uint64(len(serializedOutpoint)))
+func serializeUTXO(w io.Writer, entry *UTXOEntry, outpoint *wire.Outpoint, compressEntries bool) error {
+	err := serializeOutpoint(w, outpoint)
 	if err != nil {
 		return err
 	}
 
-	err = binary.Write(w, byteOrder, serializedOutpoint)
-	if err != nil {
-		return err
-	}
-
-	serializedUTXOEntry := serializeUTXOEntry(entry)
-	err = wire.WriteVarInt(w, uint64(len(serializedUTXOEntry)))
-	if err != nil {
-		return err
-	}
-	err = binary.Write(w, byteOrder, serializedUTXOEntry)
+	err = serializeUTXOEntry(w, entry, compressEntries)
 	if err != nil {
 		return err
 	}
 	return nil
 }
 
-// serializeUTXOEntry returns the entry serialized to a format that is suitable
-// for long-term storage. The format is described in detail above.
-func serializeUTXOEntry(entry *UTXOEntry) []byte {
+// p2pkhCompressedUTXOEntryMaxSerializeSize is the maximum serialized size for a P2PKH UTXO entry.
+// Varint (header code) + 8 bytes (amount) + compressed P2PKH script size.
+var p2pkhCompressedUTXOEntryMaxSerializeSize = wire.VarIntSerializeSize(math.MaxUint64) + 8 + cstPayToPubKeyHashLen
+
+// serializeUTXOEntry encodes the entry to the given io.Writer and use compression if useCompression is true.
+// The compression format is described in detail above.
+func serializeUTXOEntry(w io.Writer, entry *UTXOEntry, useCompression bool) error {
 	// Encode the header code.
 	headerCode := utxoEntryHeaderCode(entry)
 
-	// Calculate the size needed to serialize the entry.
-	size := serializeSizeVLQ(headerCode) +
-		compressedTxOutSize(uint64(entry.Amount()), entry.ScriptPubKey())
+	err := wire.WriteVarIntLittleEndian(w, headerCode)
+	if err != nil {
+		return err
+	}
 
-	// Serialize the header code followed by the compressed unspent
-	// transaction output.
-	serialized := make([]byte, size)
-	offset := putVLQ(serialized, headerCode)
-	offset += putCompressedTxOut(serialized[offset:], uint64(entry.Amount()),
-		entry.ScriptPubKey())
+	if useCompression {
+		return putCompressedTxOut(w, entry.Amount(), entry.ScriptPubKey())
```

Comment:
Please measure how much this compression actually saves us. With vs. without for ~finality utxoentries.
If it's less than 100mb then consider getting rid of it.